examples/plot_make_derived_metric.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) Microsoft Corporation and Fairlearn contributors.
+# Licensed under the MIT License.
+
+"""
+======================
+Making Derived Metrics
+======================
+"""
+# %%
+# This notebook demonstrates the use of the :func:`fairlearn.metrics.make_derived_metric`
+# function.
+# Many higher-order machine learning algorithms (such as hyperparameter tuners) make use
+# of scalar metrics when deciding how to proceed.
+# While the :class:`fairlearn.metrics.MetricFrame` has the ability to produce such
+# scalars through its aggregation functions, its API does not conform to that usually
+# expected by these algorithms.
+# The :func:`fairlearn.metrics.make_derived_metric` function exists to bridge this gap.
+#
+# Getting the Data
+# ================
+#
+# *This section may be skipped. It simply creates a dataset for
+# illustrative purposes*
+#
+# We will use the well-known UCI 'Adult' dataset as the basis of this
+# demonstration. This is not for a lending scenario, but we will regard
+# it as one for the purposes of this example. We will use the existing
+# 'race' and 'sex' columns (trimming the former to three unique values),
+# and manufacture credit score bands and loan sizes from other columns.
+# We start with some uncontroversial `import` statements:
+
+from fairlearn.metrics import MetricFrame, make_derived_metric
+from fairlearn.metrics import accuracy_score_group_min
+import sklearn.metrics as skm
+import functools
+import numpy as np
+import pandas as pd
+
+from sklearn.datasets import fetch_openml
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import LabelEncoder, StandardScaler
+
+
+# %%
+# Next, we import the data:
+
+data = fetch_openml(data_id=1590, as_frame=True)
+X_raw = data.data
+Y = (data.target == '>50K') * 1
+A = X_raw[['race', 'sex']]
+
+# %%
+# With the data imported, we perform some standard processing, and a test/train split:
+le = LabelEncoder()
+Y = le.fit_transform(Y)
+
+le = LabelEncoder()
+
+sc = StandardScaler()
+X_dummies = pd.get_dummies(X_raw)
+X_scaled = sc.fit_transform(X_dummies)
+X_scaled = pd.DataFrame(X_scaled, columns=X_dummies.columns)
+
+X_train, X_test, Y_train, Y_test, A_train, A_test = train_test_split(X_scaled, Y, A,
+                                                                     test_size=0.3,
+                                                                     random_state=12345,
+                                                                     stratify=Y)
+
+# Ensure indices are aligned
```

Comment:
here

Reply:
I've expanded the comment in a new branch. Although it's still not terribly clear - mainly because I'm not completely clear on what happened. AFAICT `test_train_split()` really splits a numpy array, but for Pandas, it provides a 'list of indices to use'. The practical effect of this is that until these calls, `A_train[0]` (for example) may not exist, because that's in the `A_test` slice.